### PR TITLE
[action] add TxType to handle new tx types

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
add a new field `txType` in the protobuf
1.  `txType = 0`, all existing tx (before hard-fork) will be stored into `AntiqueTx` struct
2. `txType = 1, 2, 3, 4` will used (after hard-fork) to store new transactions

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/builder.go [hunk 1]
```diff
@@ -84,6 +84,11 @@ type EnvelopeBuilder struct {
 	ab  AbstractAction
 }
 
+func (b *EnvelopeBuilder) SetTxType(v uint32) *EnvelopeBuilder {
+	b.ab.txType = v
+	return b
+}
```
Comment (CoderZhi) [inline action/builder.go:90]: as chenchen suggested in previous pr, action builders by types may be a better way.

File: action/action.go [hunk 2]
```diff
@@ -17,7 +17,7 @@ import (
 )
 
 const (
-	AntiqueTxType    = 0 // when we first enable web3 tx format, version = 0 is used, all such tx are legacy tx
+	AntiqueTxType    = 0
```
Comment (CoderZhi) [inline action/action.go:20]: since the txType is totally new field, do we still need AntiqueTx? or we can start with `LegacyTxType = 0`?

**Review by CoderZhi — COMMENTED** [review]
fix unit test, the root cause may be https://github.com/iotexproject/iotex-core/blob/master/blockchain/block/testing.go#L160

File: action/builder.go (hunk 1 above)
Comment (dustinxie) [reply action/builder.go:90]: can do in another PR since changes will be quite a lot

File: action/action.go (hunk 2 above)
Comment (dustinxie) [reply action/action.go]: see 2nd commit

File: action/action.go [hunk 3]
```diff
@@ -17,11 +17,10 @@ import (
...
 const (
-	AntiqueTxType    = 0
-	LegacyTxType     = 1
-	AccessListTxType = 2
-	DynamicFeeTxType = 3
-	BlobTxType       = 4
+	LegacyTxType     = 0
+	AccessListTxType = 1
+	DynamicFeeTxType = 2
+	BlobTxType       = 3
```
Comment (dustinxie) [inline action/action.go:23]: value is same as Ethereum's type

File: action/protocol/execution/testdata-london/datacopy-accesslist.json [hunk 4]
```diff
@@ -44,9 +44,9 @@
         "expectedStatus": 1,
         "comment": "the data of return is [0x11, 0x22, 0x33]",
         "expectedBlockInfos" : {
-            "txRootHash" : "4435fd5837f8c1a506b432ff599e5187b4bea5095e1cf16eab0dc8d647986ddf",
+            "txRootHash" : "f89594db43d4964859494a835f65c411378e2e92832dbd046b958ed36ece68a5",
```
Comment (dustinxie) [inline action/protocol/execution/testdata-london/datacopy-accesslist.json:47]: b/c `AccessTxType` changes from 2 to 1, affecting tx's hash

File: action/tx_blob_test.go [hunk 5]
```diff
@@ -50,7 +50,7 @@ func TestBlobTx(t *testing.T) {
 		r.Zero(epb.Version)
 		b := MustNoErrorV(proto.Marshal(epb))
 		h := hash.Hash256b(b[:])
-		r.Equal("10da9eef18487286f7f0324069640622cb3d18e3bde75f3c487e235df9a28fac", hex.EncodeToString(h[:]))
+		r.Equal("9801a7f9829dc76de51f54896884ad439bd25420bb5cd2771995312257ec3885", hex.EncodeToString(h[:]))
```
Comment (dustinxie) [inline action/tx_blob_test.go:53]: b/c `BlobTxType` changes from 4 to 3, affecting tx's hash

File: action/tx_legacy.go [hunk 6]
```diff
@@ -19,22 +19,24 @@ var _ TxCommonInternal = (*LegacyTx)(nil)
...
 	chainID  uint32
+	version  uint32
 	nonce    uint64
 	gasLimit uint64
 	gasPrice *big.Int
 }
 
 // NewLegacyTx creates a new legacy transaction
-func NewLegacyTx(chainID uint32, nonce uint64, gasLimit uint64, gasPrice *big.Int) *LegacyTx {
+func NewLegacyTx(chainID, version uint32, nonce uint64, gasLimit uint64, gasPrice *big.Int) *LegacyTx {
```
Comment (envestcc) [inline action/tx_legacy.go:29]: The `version` parameter should not be needed in `NewLegacyTx`; the versions we create should all be version=0, and versions not equal to 0 should be set through `fromProto`.

File: action/actctx.go [hunk 7]
```diff
@@ -220,6 +208,19 @@ func (act *AbstractAction) convertToTx() TxCommonInternal {
 			blob:       act.blobData,
 		}
 	default:
-		panic(fmt.Sprintf("unsupported action version = %d", act.version))
+		panic(fmt.Sprintf("unsupported action type = %d", act.txType))
 	}
 }
+
+func (act *AbstractAction) validateTx() error {
```
Comment (envestcc) [inline action/actctx.go:215]: They are redundant validates, already checked when convert to envelope or envelope's sanitycheck

**Review by envestcc — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

File: action/tx_legacy.go (hunk 6 above)
Comment (dustinxie) [reply action/tx_legacy.go:29]: good point

File: action/actctx.go (hunk 7 above)
Comment (dustinxie) [reply action/actctx.go:215]: this is the validate for `EnvelopeBuilder` now, once we change builder into tx-specific builder, this won't be needed

**Pull request merged**